### ISSUE TITLE
Verifying data is extremely slow

### ISSUE THREAD

**Issue by jianlianggao** [transmission/transmission#1647]
Hi transmission Team and experts,

I am using transmission container image (build: 3.00-r2-ls81) to host a torrent seed. The torrent contains about 7TB of data. The verifying data process completed 10% in the first day, and another 7% in the second day. There were only 17% of data verified after 2 days. I also run one instance on a VM with transmission ( build: 3.00-r2-ls80). The verifying only completed by 30% after 10 days. How can I speed it up?

Thank you in advance for any help.

Best regards,
J

**Comment by UltimateByte** [comment]
Hi,
7TB Data... well, since all the data has to be read, if the server has activity, it can take a while. This 100% depends on your hardware.
Issue shall be closed to leverage time for devs to work on actual program issue.

**Comment by GitHubGeek** [comment]
Simliar to #1330

**Comment by jianlianggao** [comment]
> Hi,
> 7TB Data... well, since all the data has to be read, if the server has activity, it can take a while. This 100% depends on your hardware.
> Issue shall be closed to leverage time for devs to work on actual program issue.

Hi, Thank you. I am not sure which part (maybe the VM, maybe the connection between VM and the file server...) is the bottleneck. I will split the the file into 7 parts of 1 TB and rebuild torrent files.

**Comment by UltimateByte** [comment]
Hi @jianlianggao ,
Usually the IO speed (storage drives speed) will be the bottleneck. On very low end CPUs, hashing (verifying) a file/torrent might be limited by the CPU, which you can check with `top` or `htop` commands (and watch, if the process is 100% of a CPU thread, then it might be a monothreaded process), but I don't think i've ever observed that, even on cheap VPSes.

You can get an idea of your IO and CPU utilization installing Netdata (for Linux only, I assume we're talking about Linux here) and having a look at "iowait" which is the percentage of time the CPU is waiting on the disks to complete its tasks (and that Netdata shows as CPU usage on the main graph, in a pinkish color). Netdata also provides you with useful IO metrics such as "Backlog" for a drive, which is how much time the IO is late completing tasks it was ordered to complete (might go up several seconds on intensive tasks, probably because of buffering); or the "Disk utilization" which show the time the drive is busy (reported by kernel and/or disk firmware, so results may varry, especially in the case of virtual machines, but still this is usually a great indicator).
If using a container, you would install netdata on the host.

If you want an example, you can have a look at my web server's netdata https://netdata.terageek.org (note that it is HTTPS and not HTTP on port :19999 because i've setup Apache to be a reverse proxy of the original 1999 port, and have certificates and stuff).
Project is on Github as well: https://github.com/netdata/netdata/wiki
Install guide is here: https://learn.netdata.cloud/docs/agent/packaging/installer (I advise the curl kickstart.sh method which usually works quite well on basic server installs).

But before that, maybe you don't need that Netdata tool.
You can also see iowait just using the command: `top`, people don't know that, but it's the metric noted "wa" in the table, given in percents. Example :
`%Cpu(s):  0.0 us,  2.5 sy,  0.2 ni, 97.1 id,  0.1 wa,  0.0 hi,  0.0 si,  0.0 st`

You can also check the `iotop` command that can give you useful insight. I love using `iotop -oaP` when loads of processes are on a server, to have an idea of the evolutio over time of IO utilization of processes.

Ultimately, there is a slightly tricky thing to know when measuring IO: high CPU usage of non IO limited tasks will drop total iowait percentage percieved by the system, but one task might still be waiting on IO. So storage IO usage is better measured watching disk utilization in Netdata. So in other words, iowait of a task should be deduced from total CPU usage VS CPU usage of the task you think is limited by IO. For example, say a task is 99% limited by IO, and another task is using 100% of a CPU core, then system IOWait will be around 50%. That's a bit of a ticky measurement to make, so if you're discovering this, try to just get the server to do just the one tak you're trying to measure iowait on, that way the percentage will be more accurate and easier to interpret.

Hope that helps you pinpoint the bottleneck of your server ;)

**Comment by jianlianggao** [comment]
Hi @UltimateByte,

Thank you very much for your detailed explanation of observing the parameters from server side. I currently have 0.7 us and 0.0 wa on my VM which hosting the transmission-daemon. In my case, the bottleneck comes from the IO between VM and the file server, right? Anyway, since I split the huge dataset into multiple volumes, the verifying speed has been largely improved. But I got another issue, I will open it with a separate topic.

Thank you very much again for your help.

Best wishes,
J

**Comment by UltimateByte** [comment]
Hi @jianlianggao 
0.7 load would mean you have around 70% of a CPU core used (CPU usage for the process would me more relevant & accurate in this case), which could be plausible.
Though, 0.0 iowait is weird. From my testing, either with files hosted on an SMB or NFS share, it generates iowait when limited by disk or network, especially upon writing torrents. Maybe, if using a VPS, the kernel doesn't report iowait the same way (VPS have a weird way of handling kernel reporting...). In that case, check iowait on the storage server, or check network usage, might be saturated (if it is, then it goes as fast as it can already).
Seems like you got all the keys to sort this out now.

Thanks, best wishes too.
